hub/core/dataset.py

```diff
@@ -204,18 +214,53 @@ def create_tensor(
             htype=htype,
             dtype=dtype,
             sample_compression=sample_compression,
+            hash_samples=hash_samples,
             **meta_kwargs,
         )
-        self.meta.tensors.append(name)
+        
         self.storage.maybe_flush()
         tensor = Tensor(name, self.storage)  # type: ignore
-
-        self.tensors[name] = tensor
+        
+        if hash_samples:
+            hashes_tensor = self.create_tensor(HASHES_TENSOR_FOLDER, htype="hashes")
+            self._link_tensor(tensor, hashes_tensor)
+           
+        # Hashes tensor is considered as a "hidden tensor"
+        if tensor.meta.htype == "hashes":
+            self.meta.hidden_tensors.append(name)
+            self.hidden_tensors[name] = tensor
+        else:   
+            self.meta.tensors.append(name)
+            self.tensors[name] = tensor
 
         tensor.info.update(info_kwargs)
 
         return tensor
 
+    def _link_tensor(self, src: "Tensor", dest: "Tensor"):
+        """Linking source and destination tensor. The destination tensor will be set as a linked_tensor and
+
+        Args:
+            src (Union[str, Tensor]): Source tensor.
+            dest (Union[str, Tensor]): Linked tensor. Tensor being liked to source tensor. A sample appended to the source tensor
+                                        will also be appended to its linked tensor after applying a certain function (e.g hashing).
+
+        Raises:
+            TensorAlreadyLinkedError: A linked tensor can't be linked again i.e linked tensors have a many to one relationship with source tensors.
+            NotImplementedError: Source tensor can't have more than one links.
+        """
+
+        if src.meta.links:
+            raise NotImplementedError(
+                "Source tensor already contains a link to another tensor. Linking a tensor to two or more tensors is not possible."
+            )
+
+        if dest.meta.linked_tensor or src.meta.linked_tensor:
+            raise TensorAlreadyLinkedError
+
+        src.meta.links = dest.key
```

Comment:
i know you aren't supporting multiple links, but can we use a list here? in case we do allow this later. will make backwards compatibility easier

```suggestion
        src.meta.links = [dest.key]
```

Reply:
@mccrearyd Wouldn't it be confusing to have both `meta.linked_tensor` and `meta.linked_tensors`? (The first is a boolean value, second is a list of tensors.)